<h1 id="proteinortho">Proteinortho</h1>
<p>Proteinortho is a tool to detect orthologous genes within different species.</p>
<p><strong>Input</strong>: Multiple fasta files (orange boxes) with many proteins/genes (circles). 
<strong>Output</strong>: Groups (*.proteinortho) and pairs (*.proteinortho-graph) of orthologs proteins/genes.</p>
<p>For doing so, it compares similarities of given gene sequences and clusters them to find significant groups. 
The algorithm was designed to handle large-scale data and can be applied to hundreds of species at one. 
Details can be found in (<a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-12-124">doi:10.1186/1471-2105-12-124</a>).
To enhance the prediction accuracy, the relative order of genes (synteny) can be used as additional feature for the discrimination of orthologs. The corresponding extension, namely PoFF (<a href="https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0105015">doi:10.1371/journal.pone.0105015</a>), is already build in Proteinortho. The general workflow of proteinortho: </p>
<p><img src="https://www.uni-marburg.de/de/fb16/ipc/ag-lechner/graph.png/@@images/image/unimr_lead_image_sd" alt="proteinortho.workflow.png" height="250"></p>
<p>First an initial all vs. all comparison between all proteins of all species is performed to determine protein similarities (upper right image). <br>
The second stage is the clustering of similar genes to meaningful co-orthologous groups (lower right image). <br>
Connected components within this graph can be considered as putative co-orthologous groups in theory and are returned in the output (lower left image).</p>
<h1 id="new-features-of-proteinortho-version-6">New Features of Proteinortho Version 6</h1>
<ul>
<li>Implementation of various Blast alternatives for step (for -step=2 the -p= options): Diamond, MMseqs2, Last, Topaz, Rapsearch2, Blat, Ublast and Usearch</li>
<li>Multithreading support for the clustering step (-step=3)</li>
<li>Integration of the LAPACK Fortran Library for a faster clustering step (-step=3)</li>
<li>Integration of the bitscore weights in the connectivity calculation for more data dependant splits (-step=3)</li>
<li><p>Continuous Integration &amp; Continuous Development <a href="https://gitlab.com/paulklemm_PHD/proteinortho/pipelines"><img src="https://gitlab.com/paulklemm_PHD/proteinortho/badges/master/pipeline.svg" alt="pipeline status"></a> 
<details></p>
<summary>New minor features: (Click to expand)</summary>
</li>
<li><p>Output now supports OrthoXML (-xml) and HTML.</p>
</li>
<li><a href="https://gitlab.com/paulklemm_PHD/proteinortho/wikis/Tools%20and%20additional%20programs">proteinortho_history.pl</a> a new tool for tracking proteins (or pairs of proteins) in the workflow of proteinortho.</li>
<li><a href="https://gitlab.com/paulklemm_PHD/proteinortho/wikis/Tools%20and%20additional%20programs">proteinortho_summary.pl</a></li>
<li>Various test routines (make test).</li>
<li><p>New heuristics for connectivity calculation (-step=3).
</details><details></p>
<summary>6.0.12: (Click to expand)</summary>
</li>
<li><p>improved <a href="https://gitlab.com/paulklemm_PHD/proteinortho/wikis/Tools%20and%20additional%20programs">proteinortho_history.pl</a> : now the program is &quot;smarter&quot; in detecting files automatically</p>
</li>
<li>added <a href="https://gitlab.com/paulklemm_PHD/proteinortho/wikis/Tools%20and%20additional%20programs">proteinortho_summary.pl</a> : a tool for summarizing the proteinortho-graph on species level. With the output it is easy to identify weak connected species.   </li>
<li><p>removed the diamond spam
</details>
<details><summary>6.0.13: (Click to expand)</summary></p>
</li>
<li><p>added -p=autoblast : this option alows the comparison of aminoacid and nucleotide sequences. E.g. Proteom-vs-Genome: find the protein that corresponds to a given gene (/cluster). </p>
</li>
<li>added -isoform={ncbi,uniprot,trinity} option : The reciprocal best hit graph is build using isoform information (isoforms are treated equivalent). <a href="https://gitlab.com/paulklemm_PHD/proteinortho/-/wikis/FAQ#how-does-the-isoform-work">more information about --isoform</a>
</details></li>
</ul>
<p>6.0.14 : public release to <a href="https://usegalaxy.eu/">https://usegalaxy.eu/</a></p>
<p>A more detailed list of all changes: <a href="https://gitlab.com/paulklemm_PHD/proteinortho/blob/master/CHANGELOG">CHANGELOG</a></p>
<h1 id="table-of-contents">Table of Contents</h1>
<ol>
<li><a href="#installation">Installation</a></li>
<li><a href="#synopsis">Synopsis and Description</a></li>
<li><a href="#options">Options/Parameters</a></li>
<li><a href="#poff">PoFF synteny extension</a></li>
<li><a href="#output">Output description</a></li>
<li><a href="#examples">Examples</a></li>
</ol>
<h1 id="-proteinortho-wiki-https-gitlab-com-paulklemm_phd-proteinortho-wikis-table-of-contents"><a href="https://gitlab.com/paulklemm_PHD/proteinortho/wikis/">Proteinortho-Wiki</a> Table of Contents</h1>
<ol>
<li><a href="https://gitlab.com/paulklemm_PHD/proteinortho/wikis/Tools%20and%20additional%20programs">Tools and additional programs</a></li>
<li><a href="https://gitlab.com/paulklemm_PHD/proteinortho/wikis/Error-Codes">Error Codes and Troubleshooting</a> &lt;- look here if you cannot compile/run proteinortho</li>
<li><a href="https://gitlab.com/paulklemm_PHD/proteinortho/wikis/Large-compute-jobs-(the--jobs-option">Large compute jobs example</a>)</li>
<li><a href="https://gitlab.com/paulklemm_PHD/proteinortho/-/wikis/synteny-example">Synteny + Core Proteome Example</a></li>
<li><a href="https://gitlab.com/paulklemm_PHD/proteinortho/wikis/FAQ">FAQ</a> <br>
<a href="https://gitlab.com/paulklemm_PHD/proteinortho/wikis/">(...)</a></li>
</ol>
<p>Bug reports: Please have a look at chapter <a href="https://gitlab.com/paulklemm_PHD/proteinortho/wikis/Error-Codes">2.</a> first or send a mail to incoming+paulklemm-phd-proteinortho-7278443-issue-@incoming.gitlab.com. (please include the &#39;parameter-vector&#39; that is printed for all errors)
You can also send mails to lechner@staff.uni-marburg.de. Any suggestions, feedback and comments are welcome!</p>
<h1 id="installation">Installation</h1>
<p> <strong>Proteinortho comes with precompiled binaries of all executables (Linux/x86) so you should be able to run perl proteinortho6.pl in the downloaded directory for Linux/x86.</strong>
You could also move all executables to a local bin directory (e.g. with make install PREFIX=~/bin).
If you cannot execute the src/BUILD/Linux_x86_64/proteinortho_clustering, then you have to recompile with make, see the section 2. Building and installing proteinortho from source.</p>
<p><br></p>
<h4 id="installation-with-bio-conda-for-linux-osx-install-with-bioconda-https-img-shields-io-badge-install-20with-bioconda-brightgreen-svg-style-flat-http-bioconda-github-io-recipes-proteinortho-readme-html-alt-https-img-shields-io-conda-dn-bioconda-proteinortho-svg-style-flat-https-bioconda-github-io-recipes-proteinortho-readme-html-">Installation with (bio)conda (for Linux + OSX) <a href="http://bioconda.github.io/recipes/proteinortho/README.html"><img src="https://img.shields.io/badge/install%20with-bioconda-brightgreen.svg?style=flat" alt="install with bioconda"></a> <a href="https://bioconda.github.io/recipes/proteinortho/README.html"><img src="https://img.shields.io/conda/dn/bioconda/proteinortho.svg?style=flat" alt="alt"></a></h4>
<pre><code>conda <span class="hljs-keyword">install </span>-c <span class="hljs-keyword">bioconda </span>proteinortho
</code></pre><p>If you need conda (see <a href="https://docs.anaconda.com/anaconda/install/">here</a>) and the bioconda channel: <code>conda config --add channels bioconda</code>.</p>
<p><br></p>
<h4 id="installation-with-brew-for-osx-install-with-brew-https-img-shields-io-badge-install-20with-brew-brightgreen-svg-style-flat-https-formulae-brew-sh-formula-proteinortho-dl-https-img-shields-io-badge-dynamic-json-svg-label-downloads-query-27analytics-27-27install-27-27365d-27-27proteinortho-27-url-https-3a-2f-2fformulae-brew-sh-2fapi-2fformula-2fproteinortho-json-color-green-https-formulae-brew-sh-formula-proteinortho-">Installation with brew (for OSX) <a href="https://formulae.brew.sh/formula/proteinortho"><img src="https://img.shields.io/badge/install%20with-brew-brightgreen.svg?style=flat" alt="install with brew"></a> <a href="https://formulae.brew.sh/formula/proteinortho"><img src="https://img.shields.io/badge/dynamic/json.svg?label=downloads&amp;query=$[%27analytics%27][%27install%27][%27365d%27][%27proteinortho%27]&amp;url=https%3A%2F%2Fformulae.brew.sh%2Fapi%2Fformula%2Fproteinortho.json&amp;color=green" alt="dl"></a></h4>
<pre><code><span class="hljs-keyword">brew </span><span class="hljs-keyword">install </span>proteinortho
</code></pre><p>If you need brew (see <a href="https://brew.sh/index_de">here</a>)</p>
<p><br></p>
<h4 id="deploy-with-docker-install-with-docker-https-img-shields-io-badge-install-20with-docker-brightgreen-svg-style-flat-https-quay-io-repository-biocontainers-proteinortho-">Deploy with docker <a href="https://quay.io/repository/biocontainers/proteinortho"><img src="https://img.shields.io/badge/install%20with-docker-brightgreen.svg?style=flat" alt="install with docker"></a></h4>
<pre><code>docker pull quay.io<span class="hljs-regexp">/biocontainers/</span><span class="hljs-string">proteinortho:</span>TAG
</code></pre><p>you can find the TAG <a href="https://quay.io/repository/biocontainers/proteinortho?tab=tags">here</a> (e.g. 6.0.23--hfd40d39_0).</p>
<details>
  <summary>how to docker (Click to expand)</summary>

  <br>

  First define a <a href="https://quay.io/repository/biocontainers/proteinortho?tab=tags">TAG</a> with:

  <code>export TAG=&#39;put the version tag here&#39;</code>

  To start a simple bash shell with the proteinortho container use:

  <code>docker run --rm -it quay.io/biocontainers/proteinortho:$TAG bash</code>

  Here you can start/use proteinortho.
  You can change &quot;6.0.22--hfd40d39_0&quot; with any tag/version that is available <a href="https://quay.io/repository/biocontainers/proteinortho?tab=tags">here</a>. Sadly there is no &quot;:latest&quot; tag available ...

  ### Now lets try to mount your home in the proteinortho container

  This is neccessary if you want to access your local files:

  <code>docker run --rm --mount &quot;type=bind,src=/home/$(id -un),dst=/home/$(id -un)&quot; -u $(id -u):$(id -g) -it quay.io/biocontainers/proteinortho:$TAG bash</code>

  now you have your home directory mounted to /home/YOURNAME. (load your bashrc within the container : <code>source /home/YOURNAME/.bashrc</code>)


</details>

<p><br></p>
<h4 id="available-at-galaxy-europe">Available at Galaxy Europe</h4>
<p>Simply go to the european galaxy server and search for proteinortho:</p>
<pre><code><span class="hljs-symbol">https:</span><span class="hljs-comment">//usegalaxy.eu</span>
</code></pre><p>Or you can integrate proteinortho into your own galaxy instance using: <a href="https://toolshed.g2.bx.psu.edu/view/iuc/proteinortho/4850f0d15f01">proteinortho (iuc repository)</a></p>
<p><br></p>
<h4 id="installation-with-dpkg-root-privileges-are-required-">Installation with dpkg (root privileges are required)</h4>
<p>Disclamer: Be aware that this method usually lacks 6-12 months behind the latest version</p>
<p>The deb package can be downloaded here: <a href="https://packages.debian.org/unstable/proteinortho">unstable</a> or <a href="https://packages.debian.org/stable/proteinortho">stable</a>.
Afterwards the deb package can be installed with <code>sudo dpkg -i proteinortho*deb</code>.</p>
<p><br></p>
<h4 id="-installation-with-apt-get-"><em>Installation with apt-get</em></h4>
<p>Disclamer: Be aware that this method usually lacks 6-12 months behind the latest version (<a href="https://packages.debian.org/stable/proteinortho">current version</a>)</p>
<p><em>proteinortho is released to stable Debian 11 (2021), so you can install it with <code>(sudo) apt install proteinortho</code></em></p>
<p><br></p>
<h4 id="prerequisites-for-compiling-proteinortho-from-source">Prerequisites for compiling proteinortho from source</h4>
<p>Proteinortho uses standard software which is often installed already or is part of then package repositories and can thus easily be installed. The sources come with a precompiled version of Proteinortho for 64bit Linux x86.</p>
<details>
  <summary>To <b>run</b> Proteinortho, you need: (Click to expand)</summary>


   - At least one of the following the following programs (default is diamond):

     - NCBI BLAST+ or NCBI BLAST legacy (to test this, type tblastn. apt-get install ncbi-blast+)
     - Diamond (apt-get install diamond, brew install diamond, conda install diamond, <a href="https://github.com/bbuchfink/diamond">https://github.com/bbuchfink/diamond</a>)
     - Last (<a href="http://last.cbrc.jp/">http://last.cbrc.jp/</a>)
     - Rapsearch (<a href="https://github.com/zhaoyanswill/RAPSearch2">https://github.com/zhaoyanswill/RAPSearch2</a>)
     - Topaz (<a href="https://github.com/ajm/topaz">https://github.com/ajm/topaz</a>)
     - usearch (<a href="https://www.drive5.com/usearch/download.html">https://www.drive5.com/usearch/download.html</a>)
     - ublast (is part of usearch)
     - blat (<a href="http://hgdownload.soe.ucsc.edu/admin/">http://hgdownload.soe.ucsc.edu/admin/</a>)
     - mmseqs2 (conda install mmseqs2, <a href="https://github.com/soedinglab/MMseqs2">https://github.com/soedinglab/MMseqs2</a>)
   - Perl v5.08 or higher (to test this, type perl -v in the command line)
   - (optional) Python v3.0 or higher to include synteny analysis (to test this, type &#39;python -V&#39; in the command line)
   - Perl standard modules (these should come with Perl): Thread::Queue, File::Basename, Pod::Usage, threads (if you miss one just install with <code>cpan install ...</code> )
</details>

<p><br></p>
<details>
  <summary>To <b>compile</b> Proteinortho (linux/osx), you need: (Click to expand)</summary>

   - GNU make (to test this, type &#39;make&#39; in the command line)
   - GNU g++ v4.1 or higher (to test this, type &#39;g++ --version&#39; in the command line)
   - openmp (to test this, type &#39;g++ -fopenmp&#39; in the command line)
   - (optional) gfortran for compiling LAPACK (to test this, type &#39;whereis gfortran&#39; in the command line)
   - (optional) CMake for compiling LAPACK (to test this, type &#39;cmake&#39; in the command line), OR you can use your own compiled version of lapack (you can get this with &#39;apt-get install liblapack3&#39;) and run &#39;make USEPRECOMPILEDLAPACK=TRUE&#39;

</details>

<p><br></p>
<h4 id="building-and-installing-proteinortho-from-source-linux-and-osx-">Building and installing proteinortho from source (linux and osx)</h4>
<p>  You need to have a working lapack library, check this e.g. with &#39;dpkg --get-selections | grep lapack&#39;. Install lapack e.g. with &#39;apt-get install libatlas3-base&#39; or liblapack3.</p>
<p>  If you dont have Lapack, then &#39;make&#39; will automatically compiles an old lapack (v3.8.0) for you automatically as fallback !</p>
<p>  Fetch the latest source code archive downloaded from <a href="https://gitlab.com/paulklemm_PHD/proteinortho/-/archive/master/proteinortho-master.zip">here</a></p>
<details> <summary>or from here (Click to expand)</summary>

  &gt; git clone <a href="https://gitlab.com/paulklemm_PHD/proteinortho">https://gitlab.com/paulklemm_PHD/proteinortho</a>

  &gt; wget <a href="https://gitlab.com/paulklemm_PHD/proteinortho/-/archive/master/proteinortho-master.zip">https://gitlab.com/paulklemm_PHD/proteinortho/-/archive/master/proteinortho-master.zip</a>
</details>
<br>

  - <code>tar -xzvf proteinortho*.tar.gz</code> or <code>unzip proteinortho*.zip</code> : Extract the files
  - <code>cd proteinortho*</code> : Change directory into the extracted folder
  - You can now run proteinortho6.pl directly (linux only).
  - <code>make clean &amp;&amp; make</code> : If you want to recompile Proteinortho. (For osx you need a newer g++ compiler to support multithreading, see below)
  - <code>make test</code> : To make sure Proteinortho works as expected. The output should look like below (&#39;Make test output&#39;).
  - <code>make install</code> or <code>make install PREFIX=~/bin</code> if you dont have root privileges.

<details>
  <summary><b>OSX additional informations (the -fopenmp error)</b></summary>
<pre>
Install a newer g++ compiler for -fopenmp support (multithreading) with brew (get brew here <a href="https://brew.sh/index_de">https://brew.sh/index_de</a>)

<code>brew install gcc --without-multilib</code>

Then you should have a g++-7 or whatever newer version that there is (g++-8,9,...).
Next you have to tell make to use this new compiler with one of the following:
<code>ln -s /usr/local/bin/gcc-7 /usr/local/bin/gcc
ln -s /usr/local/bin/g++-7 /usr/local/bin/g++</code>

OR(!) specify the new g++ in &#39;make CXX=/usr/local/bin/g++-7 all&#39;
</pre>
</details>

<details>
  <summary>&#39;make&#39; successful output (Click to expand)</summary>
<pre>
[  0%] Prepare proteinortho_clustering ...
[ 20%] Building <strong>proteinortho_clustering</strong> with LAPACK (static/dynamic linking)
[ 25%] Building <strong>graphMinusRemovegraph</strong>
[ 50%] Building <strong>cleanupblastgraph</strong>
[ 75%] Building <strong>po_tree</strong>
[100%] Everything is compiled with no errors.
</pre>

The compilation of proteinortho_clustering has multiple fall-back routines. If everything fails please look here <a href="https://gitlab.com/paulklemm_PHD/proteinortho/wikis/Error%20Codes">Troubleshooting (proteinortho wiki)</a>.

</details>

<h4 id="make-test-output">Make test output</h4>
<details>
  <summary>&#39;make test&#39; successful output (Click to expand)</summary>
<pre>
Everything is compiled with no errors.
[TEST] 1. basic proteinortho6.pl -step=2 tests
 [1/11] -p=blastp+ test: passed
 [2/11] -p=blastp+ synteny (PoFF) test: passed
 [3/11] -p=diamond test: passed
 [4/11] -p=diamond (--moresensitive) test (subparaBlast): passed
 [5/11] -p=lastp (lastal) test: passed
 [6/11] -p=topaz test: passed
 [7/11] -p=usearch test: passed
 [8/11] -p=ublast test: passed
 [9/11] -p=rapsearch test: passed
 [10/11] -p=blatp (blat) test: passed
 [11/11] -p=mmseqsp (mmseqs) test: passed
[TEST] 2. -step=3 tests (proteinortho_clustering)
 [1/2] various test functions of proteinortho_clustering (-test): passed
 [2/2] Compare results of &#39;with lapack&#39; and &#39;without lapack&#39;: passed
[TEST] Clean up all test files...
[TEST] All tests passed
</pre>
</details>

<p>If you have problems compiling/running the program go to <a href="https://gitlab.com/paulklemm_PHD/proteinortho/wikis/Error%20Codes">Troubleshooting (proteinortho wiki)</a>.</p>
<p><br></p>
<h1 id="synopsis">SYNOPSIS</h1>
<blockquote>
<p><strong>proteinortho [options] \<fasta file(s)\></strong></p>
</blockquote>
<p>   one fasta file for each input species; at least two species are required</p>
<h1 id="description">DESCRIPTION</h1>
<p>  <strong>proteinortho</strong> is a tool to detect orthologous genes within different
  species. </p>
<p>  Proteinortho assumes, that you have all your gene sequences in FASTA
  format either represented as amino acids or as nucleotides. The source
  code archive contains some examples, namely C.faa, E.faa, L.faa, M.faa
  located in the test/ directory. <strong>By default Proteinortho assumes amino</strong>
  <strong>acids sequences and thus uses diamond</strong> (-p=diamond) to compare sequences. If you have
  nucleotide sequences, you need to change this by adding the parameter
  -p=blastn+ (or some other algorithm). (In case you have only have NCBI
  BLAST legacy installed, you need to tell this too - either by adding
  -p=blastp or -p=blastn respectively.) The full command for the example
  files would thus be</p>
<blockquote>
<p>proteinortho6.pl -project=test test/C.faa test/E.faa</p>
</blockquote>
<p>  test/L.faa test/M.faa. Instead of naming the FASTA files one by one, you
  could also use test/*.faa. Please note that the parameter
  -project=test is optional, for naming the output. With this, you can set the prefix of the output
  files generated by Proteinortho. If you skip the project parameter, the
  default project name will be myproject.</p>
<h1 id="options-graphical-user-interface">OPTIONS graphical user interface</h1>
<p>Open <code>proteinorthoHelper.html</code> in your favorite browser or visit <a href="http://lechnerlab.de/proteinortho/">lechnerlab.de/proteinortho</a> online for an interactiv exploration of the different options of proteinortho.</p>
<h1 id="options">OPTIONS</h1>
<p> <strong>Main parameters</strong> (can be used with -- or -)</p>
<ul>
<li><p><strong>--project</strong>=name (default: myproject)
prefix for all resulting file names</p>
</li>
<li><p><strong>--inproject</strong>=name (default: same as --project)
load data from this namespace instead (works with intermediate files for step=2 and blast-graph for step=3)</p>
</li>
<li><p><strong>--cpus</strong>=number (default: all available)
the number of processors to use (multicore/processor support)</p>
<ul>
<li><p><strong>--ram</strong>=number (default: 90% of free memory)
maximal used ram threshold for LAPACK and the input graph in MB</p>
</li>
<li><p><strong>--verbose</strong>={0,1,2} (default: 1)
verbose level. 1:keeps you informed about the progress</p>
</li>
<li><p><strong>--silent</strong>
sets verbose level to 0.</p>
</li>
<li><p><strong>--temp</strong>=directory(.)
path to the temporary files</p>
</li>
<li><p><strong>--force</strong>
forces the recalculation of the blast results in any case in step=2. Also forces the recreation of the database generation in step=1</p>
</li>
<li><p><strong>--clean</strong>
removes all database-index-files generated by the -p algorithm afterwards</p>
</li>
<li><p><strong>--step</strong>={0,1,2,3} (default: 0)
0 -&gt; all. 1 -&gt; prepare blast (build db). 2 -&gt; run all-versus-all
blast. 3 -&gt; run the clustering.</p>
</li>
</ul>
<details>
 <summary>(Show more information)</summary>

   proteinortho test/<em>faa 

   # the following 3 commands are producing the same results as the command above
   proteinortho -step=1 test/</em>faa 
   proteinortho -step=2 test/*faa 
   proteinortho -step=3

</details>   

<ul>
<li><strong>--keep</strong>
stores temporary blast results for reuse (proteinortho_cache_project directory). 
In a second run the intermediate blast results are loaded instead of calculated.
You can adjust the parameters e.g. a more strict -e evalue cut off and write the output to a different namespace using --inproject.</li>
</ul>
<details>
 <summary>(Show more information)</summary>

   # 1. generate db files

   proteinortho -step=1 -project=test -keep infile/<em>fasta

   # 2. run the all-versus-all blast of some input files (infile/) 

   proteinortho -step=2 -project=test -keep infile/</em>fasta

   # now you can insert more fasta files to infile/ and reuse everything computed 

   proteinortho -step=2 -project=test -keep infile/*fasta

   # finally run clustering

   proteinortho -step=3 -project=test -keep

</details>     

<ul>
<li><strong>--isoform</strong>={ncbi,uniprot,trinity} <a href="https://gitlab.com/paulklemm_PHD/proteinortho/-/wikis/FAQ#how-does-the-isoform-work">more information about --isoform</a></li>
</ul>
<p>Merge isoforms to a single entity. </p>
<details><summary>ncbi</summary> 

   isoforms are specified in ncbi style 

   ---
   &gt;<strong>ENSMUSP00000021091.8</strong> pep chromosome:GRCm38:11:74673949:74724670:-1 <strong>gene:ENSMUSG00000020745.15</strong> transcript:ENSMUST00000021091.14 gene_biotype:protein_coding transcript_biotype:protein_coding gene_symbol:Pafah1b1 description:platelet-activating factor acetylhydrolase, <strong>isoform</strong> 1b, subunit 1 [Source:MGI Symbol;Acc:MGI:109520]
   &gt;<strong>ENSMUSP00000099578.2</strong> pep chromosome:GRCm38:11:74673950:74723858:-1 <strong>gene:ENSMUSG00000020745.15</strong> transcript:ENSMUST00000102520.8 gene_biotype:protein_coding transcript_biotype:protein_coding gene_symbol:Pafah1b1 description:platelet-activating factor acetylhydrolase, <strong>isoform</strong> 1b, subunit 1 [Source:MGI Symbol;Acc:MGI:109520]<br>   ---

   Different protein identifier (ENSMUSP00000021091.8, ENSMUSP00000099578.2) but the same gene id (ENSMUSG00000020745.15). The word &#39;isoform&#39; is also mandatory!

</details>
<details><summary>uniprot</summary> 

   isoforms are specified in uniprot style using the &#39;<em>_additional.fa&#39; files 

   E.g. C.fa: 

   ---
   &gt;tr|ADHA2|R4GDP1_DANRE Gamma-aminobutyric
   (...)
   ---

   C_additional.fa: 

   ---
   &gt;tr|QDHQ4|R4GDP1_DANRE isoform of ADHA2
   (...)
   ---

   QDHQ4 is the isoform of ADHA2. Please simply add the </em>_additional.fa files to the proteinortho call!

</details>
<details><summary>trinity</summary> 

   isoforms are specified in trinity style:

   ---
   &gt;TRINITY_DN1000_c115_g5_i1 len=247 path=[31015:0-148 23018:149-246]
   (...)
   ---

   The protein id is TRINITY_DN1000_c115_g5a and the isoform id is specified with i1

</details>

<p><strong>Search options (step 1-2)</strong>
(output: <myproject>.blast-graph)</p>
<ul>
<li><strong>--p</strong>=algorithm (default: diamond) </li>
</ul>
<p><details></p>
 <summary>show all options (Click to expand)</summary>

<ul>
<li><p>autoblast : automatically detects the blast+ program (blastp,blastn,tblastn,blastx) depending on the input (can also be mixed together!)</p>
</li>
<li><p>blastn_legacy,blastp_legacy,tblastx_legacy : legacy blast family (shell commands: blastall -) family. The suffix &#39;n&#39; or &#39;p&#39; indicates nucleotide or protein input files.</p>
</li>
<li><p>blastn+,blastp+,tblastx+ : standard blast family (shell commands: blastn,blastp,tblastx)
family. The suffix &#39;n&#39; or &#39;p&#39; indicates nucleotide or protein input files.</p>
</li>
<li><p>diamond : Only for protein files! standard diamond procedure and for
genes/proteins of length &gt;40 with the additional --sensitive flag
Warning: Please use version 0.9.29 or later to avoid this known bug: #24</p>
</li>
<li><p>lastn,lastp : lastal. -n : dna files, -p protein files (BLOSUM62 scoring matrix)!</p>
</li>
<li><p>rapsearch : Only for protein files!</p>
</li>
<li><p>mmseqsp,mmseqsn : mmseqs2. -n : dna files, -p protein files</p>
</li>
<li><p>topaz : Only for protein files!</p>
</li>
<li><p>usearch : usearch_local procedure with -id 0 (minimum identity
percentage).</p>
</li>
<li><p>ublast : usearch_ublast procedure.</p>
</li>
<li><p>blatp,blatn : blat. -n : dna files, -p protein files
</details>
<br></p>
</li>
</ul>
<ul>
<li><strong>--sim</strong>=float (default: 0.95)
min. reciprocal similarity for additional hits. 1 : only the best reciprocal hits are reported, 0 : all possible reciprocal blast matches (within the -evalue) are reported.</li>
</ul>
</li>
</ul>
<details>
  <summary>More (Click to expand)</summary>

  - <strong>--e</strong>=evalue (default: 1e-05)
    E-value for blast

  - <strong>--selfblast</strong>
    apply selfblast, detects paralogs without orthologs

  - <strong>--identity</strong>=number (default: 25)
    min. percent identity of best blast hits

  - <strong>--cov</strong>=number (default: 50)
    min. coverage of best blast alignments in %

  - <strong>--subparaBlast</strong>=&#39;options&#39;
    additional parameters for the search tool (-p=blastp+,diamond,...) example -subpara=&#39;-seg no&#39;
    or -subpara=&#39;--more-sensitive&#39; for diamond

  - <strong>--identical</strong>
    only return entries that are 100% identical

    - <strong>--range</strong>=number (default:disabled)
    maximal length difference for any blast hit. e.g. 0 = filter for hits between proteins of same length

</details>
<br>

 <strong>Synteny options (optional, step 2)</strong>
  (output: <myproject>.ffadj-graph, <myproject>.poff-graph, <myproject>.poff.tsv (tab separated file with groups))

<details>
  <summary>More (Click to expand)</summary>

  - <strong>--synteny</strong>
    activate PoFF extension to separate similar by contextual adjacencies
    (requires .gff for each .fasta)

  - <strong>--dups</strong>=number (default: 0)
    PoFF: number of reiterations for adjacencies heuristic, to determine
    duplicated regions

  - <strong>--cs</strong>=number (default: 3)
    PoFF: Size of a maximum common substring (MCS) for adjacency matches

  - <strong>--alpha</strong>=number (default: .5)
    PoFF: weight of adjacencies vs. sequence similarity
</details>
<br>

 <strong>Clustering options (step 3)</strong>
  (output: <myproject>.proteinortho.tsv, <myproject>.proteinortho.html, <myproject>.proteinortho-graph)

  - <strong>--conn</strong>=float (default: 0.1)
    min. algebraic connectivity. <b>This is the main parameter for the clustering step.</b> Choose larger values then more splits are done, resulting in more and smaller clusters. (There are still cluster with an alg. conn. below this given threshold allowed if the protein to species ratio is good enough, see -minspecies option below)

<details>

  <summary>More (Click to expand)</summary>

  - <strong>--singles</strong>
    report singleton genes without any hit

  - <strong>--purity</strong>=float (default: 1e-7)
    avoid spurious graph assignments

  - <strong>--minspecies</strong>=float (default: 1, must be &gt;=0)
    min. number of genes per species. If a group is found with up to (minspecies) genes/species, it wont be split again (regardless of the connectivity).

  - <strong>--nograph</strong>
    do not generate <em>-graph file (pairwise orthology relations)

  - <strong>--subparaCluster</strong>=&#39;options&#39;
    additional parameters for the clustering algorithm (proteinortho_clustering) example -subparaCluster=&#39;-maxnodes 10000&#39;.
    Note: -rmgraph cannot be set. All other parameters of subparaCluster are replacing the default values (like -cpus or -minSpecies)

  - <strong>--xml</strong>
    do generate an orthologyXML file (see <a href="http://www.orthoxml.org">http://www.orthoxml.org</a> for more information). You can also use proteinortho2xml.pl myproject.proteinortho.

  - <strong>--core</strong>
    stop clustering if a split would result in groups that do not span at least -coreMinSpecies species. Overrules the -conn threshold.

  - <strong>--coreMinSpecies</strong>
    sets the minimal number of species for the -core option (default:0)

  - <strong>--coreMaxProts</strong>
    sets the maximal number of proteins per species for the -core option (default:100)

</details>
<br>

 <strong>Misc options</strong>

  - <strong>--checkfasta</strong>
    checks input fasta files if the given algorithm can process the given fasta file.

<details>
  <summary>(Click to expand)</summary>

  - <strong>--cleanblast</strong>
    cleans blast-graph with proteinortho_cleanupblastgraph

  - <strong>--desc</strong>
    write description files (for NCBI FASTA input only)

  - <strong>--binpath</strong>=directory (default: $PATH)
    path to your local executables (blast, diamond, mcl, ...)

  - <strong>--debug</strong>
    gives detailed information for bug tracking

</details>
<br>

 <strong>Large compute jobs</strong>
  - <strong>--jobs</strong>=M/N
    If you want to involve multiple machines or separate a Proteinortho
    run into smaller chunks, use the -jobs=<strong>M</strong>/<strong>N</strong> option. First, run
    &#39;proteinortho6.pl -steps=1 ...&#39; to generate the indices. Then you can
    run &#39;proteinortho6.pl -steps=2 -jobs=<strong>M</strong>/<strong>N</strong> ...&#39; to run small chunks
    separately. Instead of <strong>M</strong> and <strong>N</strong> numbers must be set representing the
    number of jobs you want to divide the run into (<strong>M</strong>) and the job
    division to be performed by the process. E.g. to divide a Proteinortho
    run into 4 jobs to run on several machines, use &#39;proteinortho6.pl -steps=2 -jobs=1/4&#39;, &#39;proteinortho6.pl -steps=2 -jobs=1/4&#39;, &#39;proteinortho6.pl -steps=2 -jobs=2/4&#39;, &#39;proteinortho6.pl -steps=2 -jobs=3/4&#39;, &#39;proteinortho6.pl -steps=2 -jobs=4/4&#39;.

    See <a href="https://gitlab.com/paulklemm_PHD/proteinortho/wikis/Large-compute-jobs-(the--jobs-option">Large compute jobs, the --jobs option (proteinortho wiki)</a>) for more details.

<br>

# PoFF

  The PoFF extension allows you to use the relative order of genes (synteny)
  as an additional criterion to disentangle complex co-orthology relations.
  To do so, add the parameter -synteny. You can use it to either come closer
  to one-to-one orthology relations by preferring synthetically conserved
  copies in the presence of two very similar paralogs (default), or just to
  reduce noise in the predictions by detecting multiple copies of genomic
  areas (add the parameter -dups=3). Please note that you need additional
  data to include synteny, namely the gene positions in GFF3 format.
  AsProteinortho is primarily made for proteins, it will only accept GFF
  entries of type CDS (column #3 in the GFF-file). The attributes column
  (#9) must contain Name=GENE IDENTIFIER where GENE IDENTIFIER corresponds
  to the respective identifier in the FASTA format. It may not contain a
  semicolon (;)! Alternatively, you can also set ID=GENE IDENTIFIER. Example
  files are provided in the source code archive. Hence, we can run
  proteinortho6.pl -project=test -synteny test/A1.faa test/B1.faa test/E1.faa
  test/F1.faa to add synteny information to the calculations. Of course,
  this only makes sense if species are sufficiently similar. You won&#39;t gain
  much when comparing e.g. bacteria with fungi. When the analysis is done
  you will find an additional file in your current working directory, namely
  test.poff.tsv (tab separated file). This file is equivalent to the test.proteinortho.tsv file (above) but
  can be considered more accurate as synteny was involved for its
  construction.

  A full example is described here: <a href="https://gitlab.com/paulklemm_PHD/proteinortho/-/wikis/synteny-example">Synteny example (proteinortho wiki)</a>.

# Output
 <strong>BLAST Search (step 1-2)</strong>

<details>
  <summary>myproject.blast-graph (Click to expand)</summary>

    filtered raw blast data based on adaptive reciprocal best blast
    matches (= reciprocal best match matches within a range of 95% by default) 

    A line starting with # indicates the two species that are analysed below. E.g. &#39;# M.faa L.faa&#39; tells that the next lines are for species M versus species L.

    All matches are reciprocal matches. If
    e.g. a match for M_15 L_15 is shown, L_15 M_15 exists implicitly.

    E-Values and bit scores for both directions are given behind each
    match. 

    The 4 numbers below the species (e.g. &#39;# 3.8e-124        434.9...&#39;) are representing the median values for this comparison.

      # file_a    file_b
      # a   b     evalue_ab     bitscore_ab   evalue_ba     bitscore_ba
      # E.faa     C.faa<br>      # 3.8e-124        434.9   2.8e-126        442.2
      E_11  C_11  5.9e-51 190.7   5.6e-50 187.61
      E_10  C_10  3.8e-124    434.9   2.8e-126    442.2
      ...
 </details>
 <br>

 <strong>Clustering (step 3)</strong>

<details>
  <summary>myproject.proteinortho-graph (Click to expand)</summary>

    clustered version of the myproject.blast-graph.

    Its connected components are represented in myproject.proteinortho.tsv / myproject.proteinortho.html.

    The format of myproject.blast-graph is the equivalent to the myproject.blast-graph (see above).

      # file_a    file_b
      # a   b     evalue_ab     bitscore_ab   evalue_ba     bitscore_ba
      # E.faa     C.faa
      E_10  C_10  3.8e-124    434.9   2.8e-126    442.2
      E_11  C_11  5.9e-51 190.7   5.6e-50 187.6
      ...
 </details>
 <br>

 <details>
  <summary> myproject.proteinortho.tsv (Click to expand)</summary>

    The connected components of myproject.proteinortho-graph. 

    The very first column indicates the number of species covered by this group. 
    The second column indicates the number of genes included in this group. 

    If the number of genes is bigger than the number of species, there are co-orthologs present. 

    The third column gives the algebraic connectivity of the respective group. This indicates how densely the genes are connected
    in the orthology graph that was used for clustering. 
    A connectivity of 1 indicates a perfect dense cluster with each gene beeing connected/orthologous to each
    other gene. 

    By default, Proteinortho splits each group into two more dense subgroups when the connectivity is below 0.1 (can be user defined).

    Hint: you can open this file in Excel / Numbers / Open Office.

      # Species   Genes   Alg.-Conn.    C.faa   C2.faa  E.faa   L.faa   M.faa
      2   5     0.16  </em>     <em>     </em>     L_643,L_641   M_649,M_640,M_642
      3   6     0.138   C_164,C_166,C_167,C_2   <em>     </em>     L_2   M_2
      2   4     0.489   <em>     </em>     <em>     L_645,L_647   M_644,M_646

 </details>
 <br>

<a href="https://gitlab.com/paulklemm_PHD/proteinortho/wikis/Tools-and-additional-programs#proteinortho-graphblast-graph-species-summary-table">myproject.proteinortho-graph.summary</a>

 <br>
 <details>
  <summary> myproject.proteinortho.html (Click to expand)</summary>
    The html version of the myproject.proteinortho.tsv file
 </details>
 <br>

 <strong>POFF (-synteny)</strong>

  The synteny based graph files (myproject.ffadj-graph and
  myproject.poff-graph) have two additional columns: same_strand and
  simscore. The first one indicates if two genes from a match are located at
  the same strands (1) or not (-1). The second one is an internal score
  which can be interpreted as a normalized weight ranging from 0 to 1 based
  on the respective e-values. Moreover, a second comment line is followed
  after the species lines, e.g.

    # M.faa L.faa
    # Scores: 4   39    34.000000     39.000000

  <details>
  <summary>myproject.ffadj-graph (Click to expand)</summary>

    filtered blast data based on adaptive reciprocal best blast matches
    and synteny (only if -synteny is set)

 </details>
 <br>

  <details>
  <summary>myproject.poff-graph (Click to expand)</summary>

    clustered ffadj graph. Its connected components are represented in
    myproject.poff.tsv (tab separated file) (only if -synteny is set)

 </details>
 <br>


# EXAMPLES
 <strong>Calling proteinortho</strong>
  Sequences are typically given in plain fasta format like the files in
  test/

  test/C.faa:

    &gt;C_10
    VVLCRYEIGGLAQVLDTQFDMYTNCHKMCSADSQVTYKEAANLTARVTTDRQKEPLTGGY
    HGAKLGFLGCSLLRSRDYGYPEQNFHAKTDLFALPMGDHYCGDEGSGNAYLCDFDNQYGR
    ...

   test/E.faa:

    &gt;E_10
    CVLDNYQIALLRNVLPKLFMTKNFIEGMCGGGGEENYKAMTRATAKSTTDNQNAPLSGGF
    NDGKMGTGCLPSAAKNYKYPENAVSGASNLYALIVGESYCGDENDDKAYLCDVNQYAPNV
    ...

  To run proteinortho for these sequences, simply call

    perl proteinortho6.pl test/C.faa test/E.faa test/L.faa test/M.faa

  To give the outputs the name &#39;test&#39;, call

    perl proteinortho6.pl -project=test test/</em>faa

  To use blast instead of the default diamond, call

    perl proteinortho6.pl -project=test -p=blastp+ test/<em>faa

  If installed with make install, you can also call

    proteinortho -project=test -p=blastp+ test/</em>faa


# Hints
  Using .faa to indicate that your file contains amino acids and .fna to
  show it contains nucleotides makes life much easier.

  Sequence IDs must be unique within a single FASTA file. Consider renaming
  otherwise. Note: Till version 5.15 sequences IDs had to be unique among
  the whole dataset. Proteinortho now keeps track of name and species to
  avoid the necessissity of renaming.

  You need write permissions in the directory of your FASTA files as
  Proteinortho will create blast databases. If this is not the case,
  consider using symbolic links to the FASTA files.

  The directory src contains useful tools, e.g. proteinortho_grab_proteins.pl which
  fetches protein sequences of orthologous groups from Proteinortho output
  table. (These files are installed during &#39;make install&#39;)

# Kmere Heuristic

## Example 1

In the following example a huge blast graph is used for step 3 (clustering).
The first connected component contains 7410694 nodes, hence the kmere heuristic is activated.
Since the fiedler vector would result in a good split, the kmere heuristic is then deactivated immediatly.

<details>
  <summary>as fallback (Click to expand)</summary>

    ...
    [CRITICAL WARNING]   Failed to partition subgraph with 6929 nodes into (6929,0,0) sized groups, now using kmere heuristic as fall-back.
    ...

</details>

<details>
<summary>working example for large graphs (Click to expand)</summary>

    ...
    17:32:15 [DEBUG] (kmere-heuristic) The current connected component is so large that the k-mere heuristic can be used. First: Testing if a normal split would result in a good partition (|.|&gt;20%) of the CC.
     [WARNING] (kmere-heuristic) A normal split would NOT result in a good partition (|.|&gt;20%) of the CC, therefore  the k-mere heuristic is now used. The current connected component will be split in 3.85373 (= number of proteins <6929> / ( n
    odes per species <1> * number of species <1798>)) groups greedily accordingly to the fiedler vector.
    ...

</details>

<details>
<summary>example for large graphs, where kmere is tested but not needed (Click to expand)</summary>

    ...
    20:27:07 [DEBUG] (kmere-heuristic) The current connected component is so large that the k-mere heuristic can be used. First: Testing if a normal split would result in a good partition (|.|&gt;20%) of the CC.
    20:27:09 [DEBUG] (kmere-heuristic) A normal split would result in a good partition (|.|&gt;20%) of the CC, therefore returning now to the normal algorithm (no k-mere heuristic).
    ...

</details>

<h1 id="credit-where-credit-is-due">Credit where credit is due</h1>
<ul>
<li>The all-versus-all BLAST-analysis (-step=2) is only possible with (one of) the following underlying algorithms:<ul>
<li>NCBI BLAST+ or NCBI BLAST legacy (<a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download">https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download</a>)</li>
<li>Diamond (doi:10.1038/nmeth.3176, <a href="https://github.com/bbuchfink/diamond">https://github.com/bbuchfink/diamond</a>)</li>
<li>Last (doi:10.1101/gr.113985.110, <a href="http://last.cbrc.jp/">http://last.cbrc.jp/</a>)</li>
<li>Rapsearch2 (doi:10.1093/bioinformatics/btr595, <a href="https://github.com/zhaoyanswill/RAPSearch2">https://github.com/zhaoyanswill/RAPSearch2</a>)</li>
<li>Topaz (doi:10.1186/s12859-018-2290-3, <a href="https://github.com/ajm/topaz">https://github.com/ajm/topaz</a>)</li>
<li>usearch,ublast (doi:10.1093/bioinformatics/btq461, <a href="https://www.drive5.com/usearch/download.html">https://www.drive5.com/usearch/download.html</a>)</li>
<li>blat (<a href="http://hgdownload.soe.ucsc.edu/admin/">http://hgdownload.soe.ucsc.edu/admin/</a>)</li>
<li>mmseqs2 (doi:10.1038/nbt.3988 (2017). <a href="https://github.com/soedinglab/MMseqs2">https://github.com/soedinglab/MMseqs2</a>)</li>
</ul>
</li>
<li>The clustering step (-step=3) got a huge speedup with the integration of LAPACK (Univ. of Tennessee; Univ. of California, Berkeley; Univ. of Colorado Denver; and NAG Ltd., <a href="http://www.netlib.org/lapack/">http://www.netlib.org/lapack/</a>)</li>
<li>The html output of the *proteinortho.tsv (orthology groups) is enhanced by clusterize (<a href="https://github.com/NeXTs/Clusterize.js">https://github.com/NeXTs/Clusterize.js</a>), reducing the scroll lag.</li>
</ul>
<h1 id="online-information">ONLINE INFORMATION</h1>
<p>  For download and online information, see
  <a href="https://www.bioinf.uni-leipzig.de/Software/proteinortho/">https://www.bioinf.uni-leipzig.de/Software/proteinortho/</a>
  or
  <a href="https://gitlab.com/paulklemm_PHD/proteinortho">https://gitlab.com/paulklemm_PHD/proteinortho</a></p>
<h1 id="references">REFERENCES</h1>
<p>  Lechner, M., Findeisz, S., Steiner, L., Marz, M., Stadler, P. F., &amp;
  Prohaska, S. J. (2011). Proteinortho: detection of (co-) orthologs in
  large-scale analysis. BMC bioinformatics, 12(1), 124.</p>
