<h1 id="proteinortho">Proteinortho</h1>

<p>Proteinortho is a tool to detect orthologous genes within different species. For doing so, it compares similarities of given gene sequences and clusters them to find significant groups. The algorithm was designed to handle large-scale data and can be applied to hundreds of species at one. Details can be found in <a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-12-124">Lechner et al., BMC Bioinformatics. 2011 Apr 28;12:124.</a>
To enhance the prediction accuracy, the relative order of genes (synteny) can be used as additional feature for the discrimination of orthologs. The corresponding extension, namely PoFF (manuscript in preparation), is already build in Proteinortho. The general workflow of proteinortho is depicted [<img src="https://www.dropbox.com/s/7ubl1ginn3fmf8k/proteinortho_workflow.jpg?dl=0" alt="here" />].</p>

<h1 id="newfeaturesofproteinorthoversion6">New Features of Proteinortho Version 6!</h1>

<ul>
<li><p>Implementation of various Blast alternatives for step (for -step=2 the -p= options): Diamond, MMseqs2, Last, Topaz, Rapsearch2, Blat, Ublast and Usearch</p></li>

<li><p>Multithreading support for the clustering step (-step=3)</p></li>

<li><p>Integration of the LAPACK Fortran Library for a faster clustering step (-step=3)</p></li>

<li><p>Integration of the bitscore weights in the connectivity calculation for more data dependant splits (-step=3)
<details>
<summary>Minor features: (Click to expand)</summary></p></li>

<li><p>Output now supports OrthoXML (-xml) and HTML.</p></li>

<li><p>Various test routines (make test).</p></li>

<li><p>New heuristics for connectivity calculation (-step=3).
</details></p></li>
</ul>

<h1 id="continuousintegration">Continuous Integration</h1>

<p>supports
The badge 
<a href="https://gitlab.com/paulklemm_PHD/proteinortho/commits/master"><img src="https://gitlab.com/paulklemm_PHD/proteinortho/badges/master/pipeline.svg" alt="pipeline status" /></a> indicates the current status of the continuous integration (CI) among various platforms (ubuntu, centos, debian, fedora) and GNU c++ versions (5, 6, latest)
The whole git repository gets deployed on a clean docker imager (gcc:latest,gcc:5,ubuntu:latest,fedora:latest,debian:latest,centos:latest) and compiled (make all) and tested (make test). The badge is green only if all test are passed. For more information see <a href="https://gitlab.com/paulklemm_PHD/proteinortho/wikis/Continuous%20Integration">Continuous Integration (proteinortho wiki)</a>.</p>

<h1 id="tableofcontents">Table of Contents</h1>

<ol>
<li><a href="#installation">Installation</a></li>

<li><a href="#synopsis">Synopsis and Description</a></li>

<li><a href="#options">Options/Parameters</a></li>

<li><a href="#poff">PoFF synteny extension</a></li>

<li><a href="#output">Output description</a></li>

<li><a href="#examples">Examples</a></li>

<li><a href="https://gitlab.com/paulklemm_PHD/proteinortho/wikis/Error-Codes">Error Codes and Troubleshooting</a> &lt;- look here if you cannot compile/run (proteinortho wiki)</li>

<li><a href="https://gitlab.com/paulklemm_PHD/proteinortho/wikis/Large-compute-jobs-(the--jobs-option)">Large compute jobs example</a> (proteinortho wiki)</li>

<li><a href="https://gitlab.com/paulklemm_PHD/proteinortho/wikis/biological-example">Biological example</a> (proteinortho wiki)</li>
</ol>

<p>Bug reports: See chapter 7. or send a mail to incoming+paulklemm-phd-proteinortho-7278443-issue-@incoming.gitlab.com (Please include the 'Parameter-vector' that is printed for all errors)
You can also send a mail to lechner@staff.uni-marburg.de.</p>

<h1 id="installation">Installation</h1>

<p><strong>Proteinortho comes with precompiled binaries of all executables (Linux/x86) so you should be able to run perl proteinortho6.pl in the downloaded directory.</strong>
You could also move all executables to your favorite directory (e.g. with make install PREFIX=/home/paul/bin).
If you cannot execute the src/BUILD/Linux<em>x86</em>64/proteinortho_clustering, then you have to recompile with make, see the section 2. Building and installing proteinortho from source.</p>

<p><br></p>

<h4 id="easyinstallationwithbiocondaforlinuxosx">Easy installation with (bio)conda (for Linux + OSX)</h4>

<pre><code>conda install proteinortho
</code></pre>

<p>If you need conda (see <a href="https://docs.anaconda.com/anaconda/install/">here</a>) and the bioconda channel: <code>conda config --add channels defaults &amp;&amp; conda config --add channels bioconda &amp;&amp; conda config --add channels conda-forge</code>.</p>

<p><a href="http://bioconda.github.io/recipes/proteinortho/README.html"><img src="https://img.shields.io/badge/install%20with-bioconda-brightgreen.svg?style=flat" alt="install with bioconda" /></a> <a href="https://bioconda.github.io/recipes/proteinortho/README.html"><img src="https://img.shields.io/conda/dn/bioconda/proteinortho.svg?style=flat" alt="alt" /></a></p>

<p><br> </p>

<h4 id="easyinstallationwithbrewforosx">Easy installation with brew (for OSX)</h4>

<pre><code>brew install proteinortho
</code></pre>

<p>If you need brew (see <a href="https://brew.sh/index_de">here</a>)</p>

<p><a href="https://formulae.brew.sh/formula/proteinortho"><img src="https://img.shields.io/badge/install%20with-brew-brightgreen.svg?style=flat" alt="install with brew" /></a> <a href="https://formulae.brew.sh/formula/proteinortho"><img src="https://img.shields.io/badge/dynamic/json.svg?label=downloads&amp;query=$[%27analytics%27][%27install%27][%27365d%27][%27proteinortho%27]&amp;url=https%3A%2F%2Fformulae.brew.sh%2Fapi%2Fformula%2Fproteinortho.json&amp;color=green" alt="dl" /></a></p>

<p><br></p>

<h4 id="easyinstallationwithdocker">Easy installation with docker</h4>

<pre><code>docker pull quay.io/biocontainers/proteinortho
</code></pre>

<p><a href="https://quay.io/repository/biocontainers/proteinortho"><img src="https://img.shields.io/badge/install%20with-docker-brightgreen.svg?style=flat" alt="install with docker" /></a></p>

<p><br></p>

<h4 id="easyinstallationwithdpkgrootprivilegesarerequired">Easy installation with dpkg (root privileges are required)</h4>

<p>The deb package can be downloaded here: <a href="https://packages.debian.org/unstable/proteinortho">https://packages.debian.org/unstable/proteinortho</a>.
Afterwards the deb package can be installed with <code>sudo dpkg -i proteinortho*deb</code>.</p>

<p><br></p>

<h4 id="easyinstallationwithaptget"><em>(Easy installation with apt-get)</em></h4>

<p><strong>! Disclamer: Work in progress !</strong>
<em>proteinortho will be released to stable with Debian 11 (~2021), then proteinortho can be installed with <code>sudo apt-get install proteinortho</code> (currently this installes the outdated version v5.16b)</em></p>

<p><br></p>

<h4 id="1prerequisites">1. Prerequisites</h4>

<p>Proteinortho uses standard software which is often installed already or is part of then package repositories and can thus easily be installed. The sources come with a precompiled version of Proteinortho for 64bit Linux.</p>

<p><details>
  <summary>To <b>run</b> Proteinortho, you need: (Click to expand)</summary></p>

<ul>
<li><p>At least one of the following the following programs (default is diamond):</p>

<ul>
<li>NCBI BLAST+ or NCBI BLAST legacy (to test this, type tblastn. apt-get install ncbi-blast+)</li>

<li>Diamond (apt-get install diamond, brew install diamond, conda install diamond, https://github.com/bbuchfink/diamond)</li>

<li>Last (http://last.cbrc.jp/)</li>

<li>Rapsearch (https://github.com/zhaoyanswill/RAPSearch2)</li>

<li>Topaz (https://github.com/ajm/topaz)</li>

<li>usearch (https://www.drive5.com/usearch/download.html)</li>

<li>ublast (is part of usearch)</li>

<li>blat (http://hgdownload.soe.ucsc.edu/admin/)</li>

<li>mmseqs2 (conda install mmseqs2, https://github.com/soedinglab/MMseqs2)</li></ul></li>

<li><p>Perl v5.08 or higher (to test this, type perl -v in the command line)</p></li>

<li><p>Python v2.6.0 or higher to include synteny analysis (to test this, type 'python -V' in the command line) </p></li>

<li><p>Perl standard modules (these should come with Perl): Thread::Queue, File::Basename, Pod::Usage, threads (if you miss one just install with <code>cpan install ...</code> )
</details></p></li>
</ul>

<p><br>
<details>
  <summary>To <b>compile</b> Proteinortho (linux/osx), you need: (Click to expand)</summary></p>

<ul>
<li>GNU make (to test this, type 'make' in the command line)</li>

<li>GNU g++ v4.1 or higher (to test this, type 'g++ --version' in the command line) </li>

<li>openmp (to test this, type 'g++ -fopenmp' in the command line) </li>

<li>(optional) gfortran for compiling LAPACK (to test this, type 'whereis gfortran' in the command line)</li>

<li>(optional) CMake for compiling LAPACK (to test this, type 'cmake' in the command line), OR you can use your own compiled version of lapack (you can get this with 'apt-get install liblapack3') and run 'make USEPRECOMPILEDLAPACK=TRUE'</li>
</ul>

<p></details></p>

<p><br></p>

<h4 id="2buildingandinstallingproteinorthofromsourcelinuxandosx">2. Building and installing proteinortho from source (linux and osx)</h4>

<p>Here you can use a working lapack library, check this with 'dpkg --get-selections | grep lapack'. Install lapack e.g. with 'apt-get install libatlas3-base' or liblapack3.</p>

<p>If you dont have Lapack, then 'make' will automatically compiles Lapack v3.8.0 for you !</p>

<p>Fetch the latest source code archive downloaded from <a href="https://gitlab.com/paulklemm_PHD/proteinortho/-/archive/master/proteinortho-master.zip">here</a> 
<details> <summary>or from here (Click to expand)</summary></p>

<blockquote>
  <p>git clone https://gitlab.com/paulklemm_PHD/proteinortho</p>
  
  <p>wget https://gitlab.com/paulklemm_PHD/proteinortho/-/archive/master/proteinortho-master.zip
  </details>
  <br></p>
</blockquote>

<ul>
<li><code>tar -xzvf proteinortho*.tar.gz</code> or <code>unzip proteinortho*.zip</code> : Extract the files</li>

<li><code>cd proteinortho*</code> : Change directory into the extracted folder</li>

<li>You can now run proteinortho6.pl directly (linux only).</li>

<li><code>make clean &amp;&amp; make</code> : If you want to recompile Proteinortho. (For osx you need a newer g++ compiler to support multithreading, see below)</li>

<li><code>make install</code> or <code>make install PREFIX=~/bin</code> if you dont have root privileges. </li>

<li><code>make test</code> : To make sure Proteinortho works as expected. The output should look like below (3. Make test output).</li>
</ul>

<p><details>
  <summary><b>OSX additional informations (the -fopenmp error)</b></summary></p>

<pre>
Install a newer g++ compiler for -fopenmp support (multithreading) with brew (get brew here https://brew.sh/index_de)
<pre><code>brew install gcc --without-multilib
</code></pre>

Then you should have a g++-7 or whatever newer version that there is (g++-8,9,...). 
Next you have to tell make to use this new compiler with one of the following:<pre><code>ln -s /usr/local/bin/gcc-7 /usr/local/bin/gcc
ln -s /usr/local/bin/g++-7 /usr/local/bin/g++
</code></pre>

OR(!) specify the new g++ in 'make CXX=/usr/local/bin/g++-7 all'
</pre>

<p></details></p>

<p><details>
  <summary>'make' successful output (Click to expand)</summary></p>

<pre>
[  0%] Prepare proteinortho_clustering ...
[ 20%] Building **proteinortho_clustering** with LAPACK (static/dynamic linking)
[ 25%] Building **graphMinusRemovegraph**
[ 50%] Building **cleanupblastgraph**
[ 75%] Building **po_tree**
[100%] Everything is compiled with no errors.
</pre>

<p>The compilation of proteinortho_clustering has multiple fall-back routines. If everything fails please look here <a href="https://gitlab.com/paulklemm_PHD/proteinortho/wikis/Error%20Codes">Troubleshooting (proteinortho wiki)</a>.</p>

<p></details></p>

<h4 id="3maketestoutput">3. Make test output</h4>

<p><details>
  <summary>'make test' successful output (Click to expand)</summary></p>

<pre>
Everything is compiled with no errors.
[TEST] 1. basic proteinortho6.pl -step=2 tests
 [1/11] -p=blastp+ test: passed
 [2/11] -p=blastp+ synteny (PoFF) test: passed
 [3/11] -p=diamond test: passed
 [4/11] -p=diamond (--moresensitive) test (subparaBlast): passed
 [5/11] -p=lastp (lastal) test: passed
 [6/11] -p=topaz test: passed
 [7/11] -p=usearch test: passed
 [8/11] -p=ublast test: passed
 [9/11] -p=rapsearch test: passed
 [10/11] -p=blatp (blat) test: passed
 [11/11] -p=mmseqsp (mmseqs) test: passed
[TEST] 2. -step=3 tests (proteinortho_clustering)
 [1/2] various test functions of proteinortho_clustering (-test): passed
 [2/2] Compare results of 'with lapack' and 'without lapack': passed
[TEST] Clean up all test files...
[TEST] All tests passed
</pre>

<p></details></p>

<p>If you have problems compiling/running the program go to <a href="https://gitlab.com/paulklemm_PHD/proteinortho/wikis/Error%20Codes">Troubleshooting (proteinortho wiki)</a>.</p>

<p><br></p>

<h1 id="synopsis">SYNOPSIS</h1>

<blockquote>
  <p><strong>proteinortho6.pl [options] \<fasta file(s)\></strong> (one fasta for each species, at least 2)</p>
</blockquote>

<p>OR</p>

<blockquote>
  <p><strong>proteinortho [options] \<fasta file(s)\></strong></p>
</blockquote>

<h1 id="description">DESCRIPTION</h1>

<p><strong>proteinortho</strong> is a tool to detect orthologous genes within different
  species. For doing so, it compares similarities of given gene sequences
  and clusters them to find significant groups. The algorithm was designed
  to handle large-scale data and can be applied to hundreds of species at
  one. Details can be found in Lechner et al., BMC Bioinformatics. 2011 Apr
  28;12:124. To enhance the prediction accuracy, the relative order of genes
  (synteny) can be used as additional feature for the discrimination of
  orthologs. The corresponding extension, namely PoFF (manuscript in
  preparation), is already build in Proteinortho.</p>

<p>Proteinortho assumes, that you have all your gene sequences in FASTA
  format either represented as amino acids or as nucleotides. The source
  code archive contains some examples, namely C.faa, E.faa, L.faa, M.faa
  located in the test/ directory. <strong>By default Proteinortho assumes amino</strong>
  <strong>acids sequences and thus uses diamond</strong> (-p=diamond) to compare sequences. If you have
  nucleotide sequences, you need to change this by adding the parameter
  -p=blastn+ (or some other algorithm). (In case you have only have NCBI
  BLAST legacy installed, you need to tell this too - either by adding
  -p=blastp or -p=blastn respectively.) The full command for the example
  files would thus be </p>

<blockquote>
  <p>proteinortho6.pl -project=test test/C.faa test/E.faa</p>
</blockquote>

<p>test/L.faa test/M.faa. Instead of naming the FASTA files one by one, you
  could also use test/*.faa. Please note that the parameter
  -project=test is optional, for naming the output. With this, you can set the prefix of the output
  files generated by Proteinortho. If you skip the project parameter, the
  default project name will be myproject.</p>

<h1 id="optionsgraphicaluserinterface">OPTIONS graphical user interface</h1>

<p>Open <code>proteinorthoHelper.html</code> in your favorite browser or visit <a href="http://lechnerlab.de/proteinortho/">lechnerlab.de/proteinortho</a> online for an interactiv exploration of the different options of proteinortho.</p>

<h1 id="options">OPTIONS</h1>

<p><strong>Main parameters</strong> (can be used with -- or -)</p>

<ul>
<li><p><strong>--project</strong>=name (default: myproject)
prefix for all resulting file names</p></li>

<li><p><strong>--cpus</strong>=number (default: all available)
the number of processors to use (multicore/processor support)</p>

<ul>
<li><p><strong>--ram</strong>=number (default: 90% of free memory)
maximal used ram threshold for LAPACK and the input graph in MB</p></li>

<li><p><strong>--verbose</strong>={0,1,2} (default: 1)
verbose level. 1:keeps you informed about the progress</p></li>

<li><p><strong>--silent</strong>
sets verbose level to 0.</p></li>

<li><p><strong>--temp</strong>=directory(.)
path to the temporary files</p></li>

<li><p><strong>--force</strong>
forces the recalculation of the blast results in any case in step=2. Also forces the recreation of the database generation in step=1</p></li>

<li><p><strong>--clean</strong>
removes all database-index-files generated by the -p algorithm afterwards</p></li>

<li><p><strong>--step</strong>={0,1,2,3} (default: 0)
0 -> all. 1 -> prepare blast (build db). 2 -> run all-versus-all
blast. 3 -> run the clustering.</p></li></ul>

<p><strong>Search options (step 1-2)</strong>
(output: <myproject>.blast-graph)</p></li>
</ul>

<p><details>
  <summary>(Click to expand)</summary></p>

<ul>
<li><p><strong>--p</strong>=algorithm (default: diamond)</p>

<p><details>
  <summary>show all algorithms (Click to expand)</summary></p>

<pre><code>- autoblast,blastn_legacy,blastp_legacy,tblastx_legacy : legacy blast family (shell commands: blastall -) family. The suffix 'n' or 'p' indicates nucleotide or protein input files.

- autoblast : standard blast+ family 
automatically detects: blastn,blastp,tblastx,blastx depending on the input (can also be mixed together!)

- blastn+,blastp+,tblastx+ : standard blast+ family (shell commands: blastn,blastp,tblastx)
family. The suffix 'n' or 'p' indicates nucleotide or protein input files.

- diamond : Only for protein files! standard diamond procedure and for
genes/proteins of length &gt;40 with the additional --sensitive flag

- lastn,lastp : lastal. -n : dna files, -p protein files (BLOSUM62
scoring matrix)!

- rapsearch : Only for protein files! 

- mmseqsp,mmseqsn : mmseqs2. -n : dna files, -p protein files

- topaz : Only for protein files!

- usearch : usearch_local procedure with -id 0 (minimum identity
percentage).

- ublast : usearch_ublast procedure.

- blatp,blatn : blat. -n : dna files, -p protein files
</code></pre>

<p></details>
<br></p></li>

<li><p><strong>--e</strong>=evalue (default: 1e-05)
E-value for blast</p></li>

<li><p><strong>--selfblast</strong>
apply selfblast, detects paralogs without orthologs</p></li>

<li><p><strong>--sim</strong>=float (default: 0.95)
min. similarity for additional hits</p></li>

<li><p><strong>--identity</strong>=number (default: 25)
min. percent identity of best blast hits</p></li>

<li><p><strong>--cov</strong>=number (default: 50)
min. coverage of best blast alignments in %</p></li>

<li><p><strong>--subparaBlast</strong>='options'
additional parameters for the search tool (-p=blastp+,diamond,...) example -subpara='-seg no'
or -subpara='--more-sensitive' for diamond
</details>
<br></p>

<p><strong>Synteny options (optional, step 2)</strong>
(output: <myproject>.ffadj-graph, <myproject>.poff.tsv (tab separated file)-graph)</p></li>
</ul>

<p><details>
  <summary>(Click to expand)</summary></p>

<ul>
<li><p><strong>--synteny</strong>
activate PoFF extension to separate similar by contextual adjacencies
(requires .gff for each .fasta)</p></li>

<li><p><strong>--dups</strong>=number (default: 0)
PoFF: number of reiterations for adjacencies heuristic, to determine
duplicated regions</p></li>

<li><p><strong>--cs</strong>=number (default: 3)
PoFF: Size of a maximum common substring (MCS) for adjacency matches</p></li>

<li><p><strong>--alpha</strong>=number (default: .5)
PoFF: weight of adjacencies vs. sequence similarity
</details>
<br></p>

<p><strong>Clustering options (step 3)</strong>
(output: <myproject>.proteinortho.tsv, <myproject>.proteinortho.html, <myproject>.proteinortho-graph)</p></li>
</ul>

<p><details>
  <summary>(Click to expand)</summary></p>

<ul>
<li><p><strong>--singles</strong>
report singleton genes without any hit</p></li>

<li><p><strong>--purity</strong>=float (default: 1e-7)
avoid spurious graph assignments</p></li>

<li><p><strong>--conn</strong>=float (default: 0.1)
min. algebraic connectivity. <b>This is the main parameter for the clustering step.</b> Choose larger values then more splits are done, resulting in more and smaller clusters.</p></li>

<li><p><strong>--minspecies</strong>=float (default: 1, must be >=0)
min. number of genes per species. If a group is found with up to (minspecies) genes/species, it wont be split again (regardless of the connectivity).</p></li>

<li><p><strong>--nograph</strong>
do not generate *-graph file (pairwise orthology relations)</p></li>

<li><p><strong>--subparaCluster</strong>='options'
additional parameters for the clustering algorithm (proteinortho_clustering) example -subparaCluster='-maxnodes 10000'. 
Note: -rmgraph cannot be set. All other parameters of subparaCluster are replacing the default values (like -cpus or -minSpecies)</p></li>

<li><p><strong>--xml</strong>
do generate an orthologyXML file (see http://www.orthoxml.org for more information). You can also use proteinortho2xml.pl <myproject.proteinortho>.</p></li>

<li><p><strong>--exactstep3</strong>
perform step 3 without the k-mere heuristic (much slower for huge
datasets but more precise)</p></li>

<li><p><strong>--mcl</strong>
perform the clustering without the k-mere heuristic. The k-mere heuristic is only applied for very large connected components (>1e+6 nodes) and if the algorithm would start to iteratate very slowly</p></li>
</ul>

<p></details>
<br></p>

<p><strong>Misc options</strong></p>

<p><details>
  <summary>(Click to expand)</summary></p>

<ul>
<li><p><strong>--cleanblast</strong>
cleans blast-graph with proteinortho_cleanupblastgraph</p></li>

<li><p><strong>--checkfasta</strong>
checks input fasta files if the given algorithm can process the given fasta file.</p></li>

<li><p><strong>--desc</strong>
write description files (for NCBI FASTA input only)</p></li>

<li><p><strong>--binpath</strong>=directory (default: PATH)
path to your local executables (blast, diamond, mcl, ...)</p></li>

<li><p><strong>--debug</strong>
gives detailed information for bug tracking</p></li>
</ul>

<p></details>
<br></p>

<p><strong>Large compute jobs</strong></p>

<ul>
<li><p><strong>--jobs</strong>=M/N
If you want to involve multiple machines or separate a Proteinortho
run into smaller chunks, use the -jobs=<strong>M</strong>/<strong>N</strong> option. First, run
'proteinortho6.pl -steps=1 ...' to generate the indices. Then you can
run 'proteinortho6.pl -steps=2 -jobs=<strong>M</strong>/<strong>N</strong> ...' to run small chunks
separately. Instead of <strong>M</strong> and <strong>N</strong> numbers must be set representing the
number of jobs you want to divide the run into (<strong>M</strong>) and the job
division to be performed by the process. E.g. to divide a Proteinortho
run into 4 jobs to run on several machines, use 'proteinortho6.pl -steps=2 -jobs=1/4', 'proteinortho6.pl -steps=2 -jobs=1/4', 'proteinortho6.pl -steps=2 -jobs=2/4', 'proteinortho6.pl -steps=2 -jobs=3/4', 'proteinortho6.pl -steps=2 -jobs=4/4'.</p>

<p>See <a href="https://gitlab.com/paulklemm_PHD/proteinortho/wikis/Large-compute-jobs-(the--jobs-option)">Large compute jobs, the --jobs option (proteinortho wiki)</a> for more details.</p></li>
</ul>

<p><br></p>

<h1 id="poff">PoFF</h1>

<p>The PoFF extension allows you to use the relative order of genes (synteny)
  as an additional criterion to disentangle complex co-orthology relations.
  To do so, add the parameter -synteny. You can use it to either come closer
  to one-to-one orthology relations by preferring synthetically conserved
  copies in the presence of two very similar paralogs (default), or just to
  reduce noise in the predictions by detecting multiple copies of genomic
  areas (add the parameter -dups=3). Please note that you need additional
  data to include synteny, namely the gene positions in GFF3 format. 
  AsProteinortho is primarily made for proteins, it will only accept GFF
  entries of type CDS (column #3 in the GFF-file). The attributes column
  (#9) must contain Name=GENE IDENTIFIER where GENE IDENTIFIER corresponds
  to the respective identifier in the FASTA format. It may not contain a
  semicolon (;)! Alternatively, you can also set ID=GENE IDENTIFIER. Example
  files are provided in the source code archive. Hence, we can run
  proteinortho6.pl -project=test -synteny test/A1.faa test/B1.faa test/E1.faa
  test/F1.faa to add synteny information to the calculations. Of course,
  this only makes sense if species are sufficiently similar. You won't gain
  much when comparing e.g. bacteria with fungi. When the analysis is done
  you will find an additional file in your current working directory, namely
  test.poff.tsv (tab separated file). This file is equivalent to the test.proteinortho.tsv file (above) but
  can be considered more accurate as synteny was involved for its
  construction.</p>

<h1 id="output">Output</h1>

<p><strong>BLAST Search (step 1-2)</strong></p>

<p><details>
  <summary>myproject.blast-graph (Click to expand)</summary></p>

<pre><code>filtered raw blast data based on adaptive reciprocal best blast
matches (= reciprocal best match plus all reciprocal matches within a
range of 95% by default) The first two rows are just comments
explaining the meaning of each row. Whenever a further comment line (starting
with #) follows, it indicates results comparing the two species is
about to follow. E.g. # M.faa L.faa tells that the next lines representing
results for species M and L. All matches are reciprocal matches. If
e.g. a match for M_15 L_15 is shown, L_15 M_15 exists implicitly.
E-Values and bit scores for both directions are given behind each
match.
The 4 comment numbers ('# 3.8e-124        434.9...') are representing the median values of  
evalue_ab, bitscore_ab, evalue_ba and bitscore_ba.

  # file_a    file_b
  # a   b     evalue_ab     bitscore_ab   evalue_ba     bitscore_ba 
  # E.faa     C.faa   
  # 3.8e-124        434.9   2.8e-126        442.2
  E_11  C_11  5.9e-51 190.7   5.6e-50 187.61
  E_10  C_10  3.8e-124    434.9   2.8e-126    442.2
  ...
</code></pre>

<p></details>
 <br></p>

<p><strong>Clustering (step 3)</strong></p>

<p><details>
  <summary>myproject.proteinortho-graph (Click to expand)</summary>
    clustered myproject.blast-graph. Its connected components are represented in myproject.proteinortho.tsv / myproject.proteinortho.html. The format of myproject.blast-graph is the same as the
    blast-graph (see above).</p>

<pre><code>  # file_a    file_b
  # a   b     evalue_ab     bitscore_ab   evalue_ba     bitscore_ba
  # E.faa     C.faa
  E_10  C_10  3.8e-124    434.9   2.8e-126    442.2
  E_11  C_11  5.9e-51 190.7   5.6e-50 187.6
</code></pre>

<p></details>
 <br></p>

<p><details>
  <summary> myproject.proteinortho.tsv (Click to expand)</summary>
    The connected components. The first line starting with #is a comment
    line indicating the meaning of each column for each of the following
    lines which represent an orthologous group each. The very first column
    indicates the number of species covered by this group. The second
    column indicates the number of genes included in the group. Often,
    this number will equal the number of species, meaning that there is a
    single ortholog in each species. If the number of genes is bigger than
    the number of species, there are co-orthologs present. The third
    column gives rise to the algebraic connectivity of the respective
    group. Basically, this indicates how densely the genes are connected
    in the orthology graph that was used for clustering. A connectivity of
    1 indicates a perfect dense cluster with each gene similar to each
    other gene. By default, Proteinortho splits each group into two more
    dense subgroups when the connectivity is below 0.1 (can be user defined).
    Hint: you can open this file in Excel / Numbers / Open Office.</p>

<pre><code>  # Species   Genes   Alg.-Conn.    C.faa   C2.faa  E.faa   L.faa   M.faa
  2   5     0.16  *     *     *     L_643,L_641   M_649,M_640,M_642
  3   6     0.138   C_164,C_166,C_167,C_2   *     *     L_2   M_2
  2   4     0.489   *     *     *     L_645,L_647   M_644,M_646
</code></pre>

<p></details>
 <br></p>

<p><details>
  <summary> myproject.proteinortho.html (Click to expand)</summary>
    The html version of the myproject.proteinortho.tsv file
 </details>
 <br></p>

<p><strong>POFF (-synteny)</strong></p>

<p>The synteny based graph files (myproject.ffadj-graph and
  myproject.poff.tsv (tab separated file)-graph) have two additional columns: same_strand and
  simscore. The first one indicates if two genes from a match are located at
  the same strands (1) or not (-1). The second one is an internal score
  which can be interpreted as a normalized weight ranging from 0 to 1 based
  on the respective e-values. Moreover, a second comment line is followed
  after the species lines, e.g.</p>

<pre><code># M.faa L.faa
# Scores: 4   39    34.000000     39.000000
</code></pre>

<p><details>
  <summary>myproject.ffadj-graph (Click to expand)</summary></p>

<pre><code>filtered blast data based on adaptive reciprocal best blast matches
and synteny (only if -synteny is set)
</code></pre>

<p></details>
 <br></p>

<p><details>
  <summary>myproject.poff.tsv (tab separated file)-graph (Click to expand)</summary></p>

<pre><code>clustered ffadj graph. Its connected components are represented in
myproject.poff.tsv (tab separated file) (only if -synteny is set)
</code></pre>

<p></details>
 <br></p>

<h1 id="examples">EXAMPLES</h1>

<p><strong>Calling proteinortho</strong>
  Sequences are typically given in plain fasta format like the files in
  test/</p>

<p>test/C.faa:</p>

<pre><code>&gt;C_10
VVLCRYEIGGLAQVLDTQFDMYTNCHKMCSADSQVTYKEAANLTARVTTDRQKEPLTGGY
HGAKLGFLGCSLLRSRDYGYPEQNFHAKTDLFALPMGDHYCGDEGSGNAYLCDFDNQYGR
...
</code></pre>

<p>test/E.faa:</p>

<pre><code>&gt;E_10
CVLDNYQIALLRNVLPKLFMTKNFIEGMCGGGGEENYKAMTRATAKSTTDNQNAPLSGGF
NDGKMGTGCLPSAAKNYKYPENAVSGASNLYALIVGESYCGDENDDKAYLCDVNQYAPNV
...
</code></pre>

<p>To run proteinortho for these sequences, simply call</p>

<pre><code>perl proteinortho6.pl test/C.faa test/E.faa test/L.faa test/M.faa
</code></pre>

<p>To give the outputs the name 'test', call</p>

<pre><code>perl proteinortho6.pl -project=test test/*faa
</code></pre>

<p>To use blast instead of the default diamond, call</p>

<pre><code>perl proteinortho6.pl -project=test -p=blastp+ test/*faa
</code></pre>

<p>If installed with make install, you can also call</p>

<pre><code>proteinortho -project=test -p=blastp+ test/*faa
</code></pre>

<h1 id="hints">Hints</h1>

<p>Using .faa to indicate that your file contains amino acids and .fna to
  show it contains nucleotides makes life much easier.</p>

<p>Sequence IDs must be unique within a single FASTA file. Consider renaming
  otherwise. Note: Till version 5.15 sequences IDs had to be unique among
  the whole dataset. Proteinortho now keeps track of name and species to
  avoid the necessissity of renaming.</p>

<p>You need write permissions in the directory of your FASTA files as
  Proteinortho will create blast databases. If this is not the case,
  consider using symbolic links to the FASTA files.</p>

<p>The directory src contains useful tools, e.g. proteinortho<em>grab</em>proteins.pl which
  fetches protein sequences of orthologous groups from Proteinortho output
  table. (These files are installed during 'make install')</p>

<h1 id="kmereheuristic">Kmere Heuristic</h1>

<h2 id="example1">Example 1</h2>

<p>In the following example a huge blast graph is used for step 3 (clustering). 
The first connected component contains 7410694 nodes, hence the kmere heuristic is activated.
Since the fiedler vector would result in a good split, the kmere heuristic is then deactivated immediatly.</p>

<p><details>
  <summary>as fallback (Click to expand)</summary></p>

<pre><code>...
[CRITICAL WARNING]   Failed to partition subgraph with 6929 nodes into (6929,0,0) sized groups, now using kmere heuristic as fall-back.
...
</code></pre>

<p></details></p>

<p><details>
<summary>working example for large graphs (Click to expand)</summary></p>

<pre><code>...
17:32:15 [DEBUG] (kmere-heuristic) The current connected component is so large that the k-mere heuristic can be used. First: Testing if a normal split would result in a good partition (|.|&gt;20%) of the CC.
 [WARNING] (kmere-heuristic) A normal split would NOT result in a good partition (|.|&gt;20%) of the CC, therefore  the k-mere heuristic is now used. The current connected component will be split in 3.85373 (= number of proteins &lt;6929&gt; / ( n
odes per species &lt;1&gt; * number of species &lt;1798&gt;)) groups greedily accordingly to the fiedler vector.
...
</code></pre>

<p></details></p>

<p><details>
<summary>example for large graphs, where kmere is tested but not needed (Click to expand)</summary></p>

<pre><code>...
20:27:07 [DEBUG] (kmere-heuristic) The current connected component is so large that the k-mere heuristic can be used. First: Testing if a normal split would result in a good partition (|.|&gt;20%) of the CC.
20:27:09 [DEBUG] (kmere-heuristic) A normal split would result in a good partition (|.|&gt;20%) of the CC, therefore returning now to the normal algorithm (no k-mere heuristic).
...
</code></pre>

<p></details></p>

<h1 id="creditwherecreditisdue">Credit where credit is due</h1>

<ul>
<li>The all-versus-all BLAST-analysis (-step=2) is only possible with (one of) the following underlying algorithms:


<ul>
<li>NCBI BLAST+ or NCBI BLAST legacy (https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE<em>TYPE=BlastDocs&amp;DOC</em>TYPE=Download)</li>

<li>Diamond (doi:10.1038/nmeth.3176, https://github.com/bbuchfink/diamond)</li>

<li>Last (doi:10.1101/gr.113985.110, http://last.cbrc.jp/)</li>

<li>Rapsearch2 (doi:10.1093/bioinformatics/btr595, https://github.com/zhaoyanswill/RAPSearch2)</li>

<li>Topaz (doi:10.1186/s12859-018-2290-3, https://github.com/ajm/topaz)</li>

<li>usearch,ublast (doi:10.1093/bioinformatics/btq461, https://www.drive5.com/usearch/download.html)</li>

<li>blat (http://hgdownload.soe.ucsc.edu/admin/)</li>

<li>mmseqs2 (doi:10.1038/nbt.3988 (2017). https://github.com/soedinglab/MMseqs2)</li></ul>
</li>

<li>The clustering step (-step=3) got a huge speedup with the integration of LAPACK (Univ. of Tennessee; Univ. of California, Berkeley; Univ. of Colorado Denver; and NAG Ltd., http://www.netlib.org/lapack/)</li>

<li>The html output of the *proteinortho.tsv (orthology groups) is enhanced by clusterize (https://github.com/NeXTs/Clusterize.js), reducing the scroll lag.</li>
</ul>

<h1 id="onlineinformation">ONLINE INFORMATION</h1>

<p>For download and online information, see
  <a href="https://www.bioinf.uni-leipzig.de/Software/proteinortho/">https://www.bioinf.uni-leipzig.de/Software/proteinortho/</a>
  or
  <a href="https://gitlab.com/paulklemm_PHD/proteinortho">https://gitlab.com/paulklemm_PHD/proteinortho</a></p>

<h1 id="references">REFERENCES</h1>

<p>Lechner, M., Findeisz, S., Steiner, L., Marz, M., Stadler, P. F., &amp;
  Prohaska, S. J. (2011). Proteinortho: detection of (co-) orthologs in
  large-scale analysis. BMC bioinformatics, 12(1), 124.</p>